x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java

```diff
@@ -0,0 +1,123 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.rank.textsimilarity;
+
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.license.LicenseUtils;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.retriever.RetrieverBuilder;
+import org.elasticsearch.search.retriever.RetrieverParserContext;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.XPackPlugin;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.search.rank.RankBuilder.DEFAULT_RANK_WINDOW_SIZE;
+
+/**
+ * A {@code RetrieverBuilder} for parsing and constructing a text similarity reranker retriever.
+ */
+public class TextSimilarityRankRetrieverBuilder extends RetrieverBuilder {
+
+    public static final NodeFeature TEXT_SIMILARITY_RERANKER_RETRIEVER_SUPPORTED = new NodeFeature(
+        "text_similarity_reranker_retriever_supported"
+    );
+
+    public static final ParseField RETRIEVER_FIELD = new ParseField("retriever");
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField RANK_WINDOW_SIZE_FIELD = new ParseField("rank_window_size");
+    public static final ParseField INFERENCE_ID_FIELD = new ParseField("inference_id");
+    public static final ParseField INFERENCE_TEXT_FIELD = new ParseField("inference_text");
+    public static final ParseField MIN_SCORE_FIELD = new ParseField("min_score");
+
+    public static final ObjectParser<TextSimilarityRankRetrieverBuilder, RetrieverParserContext> PARSER = new ObjectParser<>(
+        TextSimilarityRankBuilder.NAME,
+        TextSimilarityRankRetrieverBuilder::new
+    );
+
+    static {
+        PARSER.declareObject((r, v) -> r.retrieverBuilder = v, RetrieverBuilder::parseInnerRetrieverBuilder, RETRIEVER_FIELD);
+        PARSER.declareField((r, v) -> r.field = v, (p, c) -> p.text(), FIELD_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField(
+            (r, v) -> r.rankWindowSize = v,
+            (p, c) -> p.intValue(),
+            RANK_WINDOW_SIZE_FIELD,
+            ObjectParser.ValueType.INT_OR_NULL
+        );
+        PARSER.declareField((r, v) -> r.inferenceId = v, (p, c) -> p.text(), INFERENCE_ID_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.inferenceText = v, (p, c) -> p.text(), INFERENCE_TEXT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.minScore = v, (p, c) -> p.floatValue(), MIN_SCORE_FIELD, ObjectParser.ValueType.FLOAT_OR_NULL);
```

Comment:
@benwtrent I switched the parser to use `ConstructingObjectParser` (https://github.com/elastic/elasticsearch/pull/109813/commits/199822e81485859876073e18a9e8631c728a38fc). 

The reason I was using the object parser was to enable tracking of sections (following [StandardRetrieverBuilder](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/retriever/StandardRetrieverBuilder.java#L54) as an example). But since that got removed, I'm switching back.

Hopefully this addresses your comments regards to parsing.